Row group limit pruning for row groups that entirely match predicates

xudong963 opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

- Part of https://github.com/apache/datafusion/issues/18860

## Rationale for this change

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

See https://github.com/apache/datafusion/issues/18860#issuecomment-3563442093

## What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

1.  How to decide if we can do limit pruning without messing up the sql semantics.
2. Add logic to decide if a row group is fully matched, all rows in the row group are matched the predicated.
3. Use the fully matched row groups to return limit rows.

## Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
3. Serve as another way to document the expected behavior of the code


If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

Yes

## Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

No, no new configs, or API change

xudong963 commented on datafusion/datasource-parquet/src/row_group_filter.rs:
```diff
@@ -135,13 +186,56 @@ impl RowGroupAccessPlanFilter {
+                        Arc::new(NotExpr::new(Arc::clone(predicate.orig_expr())));
+                    // Simplify the NOT expression (e.g., NOT(c1 = 0) -> c1 != 0)
+                    // before building the pruning predicate
+                    let mut simplifier = PhysicalExprSimplifier::new(arrow_schema);
```
@adriangb that's why I opened the PR https://github.com/apache/datafusion/pull/18587. Finally, you made me think more and I put some logic into PhysicalExprSimplifier

xudong963 commented on datafusion/datasource-parquet/src/opener.rs:
```diff
@@ -407,8 +407,12 @@ impl FileOpener for ParquetOpener {
 
-            let mut access_plan = row_groups.build();
+            // Prune by limit
+            if let Some(limit) = limit {
```
I thought a relatively clean way when fishing, lol. Will push a commit for that.

alamb commented on datafusion/datasource-parquet/src/row_group_filter.rs:
```diff
@@ -135,13 +186,56 @@ impl RowGroupAccessPlanFilter {
+                            for (i, &original_row_group_idx) in
+                                fully_contained_candidates_original_idx.iter().enumerate()
+                            {
+                                // If the inverted predicate *also* prunes this row group (meaning inverted_values[i] is false),
```
This is a very clever idea -- basically if `not(predicate)` would prune the predicate it means that `predicate` evaluated to (known) `true` for all rows

I tried thinking of counter examples, but I couldn't come up with any and the reasoning is sound

alamb commented on datafusion/datasource-parquet/src/row_group_filter.rs:
```diff
@@ -70,6 +78,49 @@ impl RowGroupAccessPlanFilter {
+        &self.is_fully_matched
+    }
+
+    /// Prunes the access plan based on the limit and fully contained row groups.
```
I think we should explain the rationale here and leave a link to the paper -- I think you can adapt the very nice description on https://github.com/apache/datafusion/issues/18860

Basically:
1. Why is this optimization important: (can reduce the number of RowGroups needed with a known limit)
2. Why is it correct (the rationale on https://github.com/apache/datafusion/issues/18860)

alamb reviewed (commented):
Thank you very much for this PR @xudong963  

The idea is very cool,  and a nicely done PR; I have several suggestions, but I think it is very close. 

Suggestions:
1. Rename the flag to `preserve_order`
1. Make this new flag on TableScan visible in EXPLAIN plans so we have a better chance of understanding when it is being applied/not applied
3. Implement an end to end test for the behavior you want to see (specifically, a test that shows a query on a parquet file and demonstrates that only a single row group is fetched with this flag, and more than one is fetched without the flag)

In my mind, 3 is the most important

xudong963 commented on datafusion/datasource-parquet/src/row_group_filter.rs:
```diff
@@ -46,13 +48,19 @@ use parquet::{
     /// which row groups should be accessed
     access_plan: ParquetAccessPlan,
+    /// which row groups are fully contained within the pruning predicate
+    is_fully_matched: Vec<bool>,
```
Do you mean to rename `is_fully_matched` as `fully_contained`?

xudong963 commented on datafusion/datasource-parquet/src/row_group_filter.rs:
```diff
@@ -135,13 +186,56 @@ impl RowGroupAccessPlanFilter {
+                    }
+                }
+
+                // Note: this part of code shouldn't be expensive with a limited number of row groups
```
Sure

xudong963 commented:
> Thank you very much for this PR @xudong963
> 
> The idea is very cool, and a nicely done PR; I have several suggestions, but I think it is very close.
> 
> Suggestions:
> 
> 1. Rename the flag to `preserve_order`
> 2. Make this new flag on TableScan visible in EXPLAIN plans so we have a better chance of understanding when it is being applied/not applied
> 3. Implement an end to end test for the behavior you want to see (specifically, a test that shows a query on a parquet file and demonstrates that only a single row group is fetched with this flag, and more than one is fetched without the flag)
> 
> In my mind, 3 is the most important

Thanks for the review @alamb . I addressed some suggestions:
1. Renamed the flag to `preseerve_order` https://github.com/apache/datafusion/pull/18868/commits/ecbfff7768542fc79fc72e94119a06c9610d29e7
2. Avoided the second recursive call in pushdown limit and maintained a state to decide if set `preserve_order` as true. https://github.com/apache/datafusion/pull/18868/commits/e919879cbf003df9fe89b58c94dc72af29ea9f72
3. Extracted the logic that finds fully matched rg as a function: https://github.com/apache/datafusion/pull/18868/commits/647618a6c8e73540c82c12dc471370055f1840b2
4. Added the end-to-end sqllogictests for limit pruning https://github.com/apache/datafusion/pull/18868/commits/63cd8783c3b9f7bdf699f94148430c4100541056
5. Extracted the not simplifier as a new PR: https://github.com/apache/datafusion/pull/18970

Things later the PR:
1. Make a TableScanBuilder or something (as a follow on PR) and deprecate TableScan::try_new
2. Add the `preserve_order` to explain. The reason I plan to do it in another PR is that I guess it'll make lots of changes for tests, which will expand the PR.

alamb commented:
I was mentioning this PR to @crepererum and he also noted there is another interesting potential optimization when we know the predicate is true for the entire row group: we can skip evaluating the predicate for the row group entirely (as the filter can often be quite expensive itself)

@xudong963 is this something else you are contemplating? This PR likely lays the foundation for such an optimization. If you think it is worthwhile I'll file a ticket

zhuqi-lucas commented:
> I wonder how this interacts with #18817? Both seem to involve the order of scans. I'm sorry I haven't looked into detail in both myself to answer the question - I've been on vacation and barely able to keep up with all of this great work. @zhuqi-lucas any thoughts?

@adriangb I think they are two independent optimization, this PR is for the optimization for all rows in the row group are matched. And the  #18817 is target to remove the sort, and i think we need benchmark to confirm the performance improvement, i will do this.

xudong963 commented:
Rebased main and removed code in the PR https://github.com/apache/datafusion/pull/18970

xudong963 commented:
Hey @alamb, I rethought the issue:

The key is that we need to accurately identify _pure_ LIMIT queries (those without an ORDER BY clause) to safely apply Limit Pruning.

Generally, if an `ORDER BY` exists, the `LIMIT` cannot be pushed down to the Parquet level. Consequently, we often determine whether to execute Limit Pruning _based on whether a limit is set within the ParquetOpener_.

`SELECT * FROM t WHERE b > 10 ORDER BY a LIMIT 10`❌ (Pruning disabled)

`SELECT * FROM t WHERE b > 10 LIMIT 10` ✅ (Pruning enabled)

However, a **complication arises**: if the data distribution and physical ordering of table t already satisfy `ORDER BY a`, the `EnforceSorting` phase in the Physical Optimizer will _remove_ the Sort node. Subsequently, the `LIMIT` is pushed down to the `DataSource` during the `LimitPushdown` phase.

In this scenario, we cannot rely _solely_ on the presence of the `LIMIT` at the Parquet level to decide whether to prune. If we prune based on a limit that was originally associated with a removed Sort, we might _violate the required global ordering_.

**Proposed Solution**: To address this _fundamentally_, when a Sort node is removed because the distribution already matches the requirements, we should mark the resulting Limit node as ["**Order-Sensitive**."](https://github.com/apache/datafusion/pull/18868/commits/473f4a97ea1bf29c925a8b9c7702896570523d18#diff-445a93b16e352019a131011d0574c5bc2e1ec32b830c0052876bdba76855e5f5R586-R593) During the subsequent `LimitPushdown`, we can [detect this flag and set ](https://github.com/apache/datafusion/pull/18868/commits/473f4a97ea1bf29c925a8b9c7702896570523d18#diff-baa20d781df314d37cdfd47e897af90ae4aabf1aa2b892a90bf0b3bdd21e020eR253-R278)the `preserve_order` attribute to true in the ParquetOpener. This ensures that Limit Pruning is bypassed when the global order must be maintained.

By this solution, we can avoid many changes and make the code neat

adriangb commented on datafusion/physical-plan/src/limit.rs:
```diff
@@ -51,6 +51,9 @@ pub struct GlobalLimitExec {
     cache: PlanProperties,
+    /// Whether the limit is order-sensitive
+    /// Such as the child plan is a sort node, then the limit is order-sensitive
+    order_sensitive: bool,
```
Just a thought: instead of order sensitive could this store the sort order it requires? If we have the info available and are adding a field that could be more useful in ensuring correctness of the plan is preserved through transformations.

xudong963 replied:
Do you mean replace the `order_sensitive` with `required_ordering: Option<LexOrdering>`

adriangb commented:
There is a proposal in #19487 that relates to this work in the "is a container known to match all filters for all rows just from stats" sense

xudong963 commented:
> There is a proposal in #19487 that relates to this work in the "is a container known to match all filters for all rows just from stats" sense

Yeah, row group limit pruning is also from the paper [Pruning in Snowflake: Working Smarter, Not Harder](https://arxiv.org/abs/2504.11540), section 4 -- Pruning for LIMIT Queries

xudong963 commented:
Hey @alamb @adriangb, do you have time to review the PR? It would be sweet to have it in 52.0.0

xudong963 commented:
> Hi @xudong963 -- I am now back from vacation and will review this PR either later today or tomorrow

Thank you @alamb, hope you have a great vacation!

alamb commented on datafusion/datasource-parquet/src/metrics.rs:
```diff
@@ -93,6 +95,11 @@ impl ParquetFileMetrics {
             .with_type(MetricType::SUMMARY)
             .pruning_metrics("row_groups_pruned_bloom_filter", partition);
 
+        let limit_pruned_row_groups = MetricBuilder::new(metrics)
```
Can you please also document this metric in the docs here? https://github.com/apache/datafusion/blob/9e9eb944c06279a4abbfd1924b80192a2f4b4ef6/docs/source/user-guide/explain-usage.md?plain=1#L229-L228

alamb commented on datafusion/datasource-parquet/src/row_group_filter.rs:
```diff
@@ -70,6 +78,52 @@ impl RowGroupAccessPlanFilter {
+            }
+        }
+
+        if fully_matched_rows_count >= limit {
```
```suggestion
        // If we can satisfy the limit with fully matching row groups, 
        // rewrite the plan to do so
        if fully_matched_rows_count >= limit {
```

alamb commented on datafusion/datasource-parquet/src/row_group_filter.rs:
```diff
@@ -46,13 +48,19 @@ use parquet::{
 pub struct RowGroupAccessPlanFilter {
     /// which row groups should be accessed
     access_plan: ParquetAccessPlan,
+    /// Row groups where ALL rows are known to match the pruning predicate
```
```suggestion
    /// Row groups where ALL rows are known to match the pruning predicate 
    /// (the predicate does not filter any rows)
```

alamb commented on datafusion/physical-plan/src/limit.rs:
```diff
@@ -51,6 +52,9 @@ pub struct GlobalLimitExec {
     metrics: ExecutionPlanMetricsSet,
     cache: PlanProperties,
+    /// If the child plan is a sort node, after the sort node is removed during
+    /// physical optimization, we should add the required ordering to the limit node
```
I think it might also help to explain why this is needed. Something like
```suggestion
    /// Does the limit have to preserve the order if its input, and if so what is it?
    /// Some optimizations may reorder the input if no particular sort is required
```

alamb commented on datafusion/sqllogictest/test_files/limit_pruning.slt:
```diff
@@ -0,0 +1,77 @@
+set datafusion.execution.parquet.pushdown_filters = true;
+
+
+statement ok
```
Shall we credit https://arxiv.org/pdf/2504.11540 ? I love these names

alamb commented on datafusion/sqllogictest/test_files/limit_pruning.slt:
```diff
@@ -0,0 +1,77 @@
+statement ok
+set datafusion.explain.analyze_level = summary;
+
+# row_groups_pruned_statistics=4 total → 3 matched -> 1 fully matched
```
this is because Row Group 2 is entirely matched (`Nlpine*`) right?

alamb commented on datafusion/sqllogictest/test_files/limit_pruning.slt:
```diff
@@ -0,0 +1,77 @@
+  ('Nlpine Sheep', 83),
+  ('Europ. Mole', 4),
+  ('Polecat', 16),
+  ('Alpine Ibex', 97);
```
When reviewing the test results, I needed to break it up like this so the expected data was easier to verify. It might help other readers to do the same

```suggestion
CREATE TABLE t AS VALUES
-- ***** Row Group 0 *****
  ('Anow Vole', 7),
  ('Brown Bear', 133),
  ('Gray Wolf', 82),
-- ***** Row Group 1 *****
  ('Lynx', 71),
  ('Red Fox', 40),
  ('Alpine Bat', 6),
-- ***** Row Group 2 *****
  ('Nlpine Ibex', 101),
  ('Nlpine Goat', 76),
  ('Nlpine Sheep', 83),
-- ***** Row Group 3 *****
  ('Europ. Mole', 4),
  ('Polecat', 16),
  ('Alpine Ibex', 97);
```

alamb reviewed (approved):
👏  this is a great PR -- thank you @xudong963 
I went over the code and tests very carefully and I think it is really nice

I would personally suggest we merge this one into main and not into 52 to give it maximum "bake time" (testing before release) but I defer to you in that regard

xudong963 commented on datafusion/datasource-parquet/src/row_group_filter.rs:
```diff
@@ -70,6 +78,52 @@ impl RowGroupAccessPlanFilter {
+    }
+
+    /// Prunes the access plan based on the limit and fully contained row groups.
+    /// See the [description](https://github.com/apache/datafusion/issues/18860#issuecomment-3563442093)
```
good point!

xudong963 commented:
> Hi @xudong963 I'm sorry for the delay. I have also been on holiday and struggling to keep up with things. I did read the Snowflake paper... great stuff! I do plan to review this PR some time this week, it's on my shortlist. Thank you for the great work and for your patience.

Thank you @adriangb , no worries, looking forward to your thoughts!

adriangb commented on datafusion/datasource-parquet/src/row_group_filter.rs:
```diff
@@ -70,6 +79,109 @@ impl RowGroupAccessPlanFilter {
+    ///
+    /// +-----------------------------------------------------------------------+
+    /// |                            NOT MATCHING                               |
+    /// |  Partition 1                                                          |
```
I think we use the terminology `container` in DataFusion. I understand the Snowflake paper uses partition / micro partition, but I think better to stick to container in our codebase.

xudong963 replied:
Good catch, I'd like to use Row group here

adriangb commented on datafusion/datasource-parquet/src/row_group_filter.rs:
```diff
@@ -153,6 +218,68 @@ impl RowGroupAccessPlanFilter {
+
+        // Simplify the NOT expression (e.g., NOT(c1 = 0) -> c1 != 0)
+        // before building the pruning predicate
+        let simplifier = PhysicalExprSimplifier::new(arrow_schema);
```
I wonder if we could short circuit by adding a `not()` to `PruningPredicate` or something like that to take an existing `PruningPredicate` and negate it by cloning the innner expression `Arc` or something.

adriangb commented on datafusion/physical-optimizer/src/limit_pushdown.rs:
```diff
@@ -337,4 +362,37 @@ fn add_global_limit(
+
+    let Some(data_source_exec) = plan.as_any().downcast_ref::<DataSourceExec>() else {
+        return plan;
+    };
```
It saddens me every time I see a downcast. Maybe you decided not to because it's more code churn or something but could we record in the FileScanConfig if an order was pushed down into it and if so set the `preserve_order` flag? As opposed to the case where the scan has an ordering but it isn't required by upstream operators.

xudong963 replied:
@adriangb I got you, shortly summarize your thought is " the Self-Driven Promise principle: any node declaring sorted output must ensure it preserves that order."

adriangb commented on datafusion/datasource-parquet/src/row_group_filter.rs:
```diff
@@ -70,6 +79,109 @@ impl RowGroupAccessPlanFilter {
+    /// |                           FULLY MATCHING                              |
+    /// |  Partition 3                                                          |
+    /// |  +-----------------------------------+-----------------------------+  |
+    /// |  | SPECIES (min: 'A...',max: 'A...') | S (min: 76, max: 101)       |  |
```
But what about `ANotAlpine`? It would not match the filter. So not all rows match. It still starts with `A`

xudong963 replied:
@adriangb Oh, I got you, I think the pic left out some details, I'll change to min: 'Alpine', max: 'Alpine Sheep'

xudong963 commented on datafusion/datasource-parquet/src/opener.rs:
```diff
@@ -76,6 +76,8 @@ pub(super) struct ParquetOpener {
     /// Optional limit on the number of rows to read
     pub limit: Option<usize>,
+    /// If should keep the output rows in order
+    pub preserve_order: bool,
```
No, we need the pub like other fields, because in `create_file_opener`, the opener is created by Structure way:
```
ParquetOpener {
            partition_index: partition,
            projection: self.projection.clone(),
            batch_size: self
                .batch_size
                .expect("Batch size must set before creating ParquetOpener"),
            limit: base_config.limit,
            preserve_order: base_config.preserve_order,
            predicate: self.predicate.clone(),
            table_schema: self.table_schema.clone(),
            metadata_size_hint: self.metadata_size_hint,
            metrics: self.metrics().clone(),
            parquet_file_reader_factory,
            pushdown_filters: self.pushdown_filters(),
            reorder_filters: self.reorder_filters(),
            force_filter_selections: self.force_filter_selections(),
            enable_page_index: self.enable_page_index(),
            enable_bloom_filter: self.bloom_filter_on_read(),
            enable_row_group_stats_pruning: self.table_parquet_options.global.pruning,
            coerce_int96,
            #[cfg(feature = "parquet_encryption")]
            file_decryption_properties,
            expr_adapter_factory,
            #[cfg(feature = "parquet_encryption")]
            encryption_factory: self.get_encryption_factory_with_config(),
            max_predicate_cache_size: self.max_predicate_cache_size(),
            reverse_row_groups: self.reverse_row_groups,
        });
```

We could add a new method later, then see if some fields don't need the pub

adriangb commented on datafusion/datasource-parquet/src/row_group_filter.rs:
```diff
@@ -70,6 +79,109 @@ impl RowGroupAccessPlanFilter {
+    /// |  | Alpine Goat  ✅                  | 76     ✅                   |  |
+    /// |  | Alpine Sheep ✅                  | 83     ✅                   |  |
+    /// |  +-----------------------------------+-----------------------------+  |
+    /// +-----------------------------------------------------------------------+
```
Can you edit the example to truncation length 3? Length 6 is conveniently the same as the needle expression, I think it's important to show what happens when it's shorter. Also I think `col like 'foo%'` does not generate a predicate involving `like`: `predicate=species@0 LIKE Alpine%, pruning_predicate=species_null_count@2 != row_count@3 AND species_min@0 <= Alpinf AND Alpine <= species_max@1`. So in this case we'd end up with `NOT ('Alp' <= 'Alpinf' AND 'Alpine' <= 'Alq')` (those are the truncated stats).

xudong963 replied:
Thank you @adriangb, I'll address it today

xudong963 commented:
I plan to merge the PR today if there are no more comments.

Thanks for the review @alamb @adriangb!

adriangb commented:
Should we also have an issue to track using this idea / machinery to avoid running filters that we know match all rows for a file during filter pushdown?

xudong963 commented:
> Should we also have an issue to track using this idea / machinery to avoid running filters that we know match all rows for a file during filter pushdown?

Yes, this one https://github.com/apache/datafusion/issues/19028

alamb commented:
This is super exciting work -- thank you so much @xudong963 and @adriangb 

Interestingly I was just speaking about this feature with @NGA-TRAN and @gene-bordegaray (who I ran into at https://nedbday.github.io/2026/). Very cool
